Design.md

```diff
@@ -77,19 +77,37 @@ This API creates a Tenant on Vault.
             }]
     }
    ```
-1. `attach-role-policy` with s3 and iam full access to this role
+1. `attach-role-policy` with s3 and iam full access to this role using generated access key
 1. `delete-access-key` for the account.
 
 ### Create Tenant Example Activity Diagram
 ![Create Tenant](Design_Files/OSE-OSIS-Create-Tenant.png)
 
-### Query Tenants
-This API will query tenants on Vault using a `filter` parameter.
+### List Tenants
+This API will list tenants on Vault. 
 1. Vault `list-accounts` api will be called using vaultclient.
-1. Filter the result accounts using the `filter` parameter in the request
+   * This API can be called with following parameters:
+        * `offset`:  The start index of tenants to return (optional)
+        * `limit`: Maximum number of tenants to return (optional)
+    
+   **Scenarios:**
+    1. If `offset` and `limit` values are not passed in the request:
+        1. `list-accounts` api will be called with `max-items` with 1000 increments with each marker until `isTruncated` is `false` in the result. The consolidated Accounts list will be returned as Tenants list 
+    1. If only `offset` is passed and `limit` value is not passed in the request:
+       1. `list-accounts` api will be called with `max-items` with 1000 increments with each marker until `isTruncated` is `false` in the result. The sublist of the consolidated Accounts list from the provided `offset` will be returned as Tenants list
+   1. If only `limit` is passed and `offset` value is not passed in the request:
+      1. If `limit <= 1000`, `list-accounts` api will be called with `max-items` with `limit` value and the Accounts list will be returned as Tenants list.
```

Comment:
If the limit is `<= 1000` it still doesn't guarantee that the first 1000 items will all have cdTenantIdList.

For example: 

Assumption:
* `limit` is 925
* There are 10000 accounts. 

There is a high possibility that in the first 1000 accounts, only 900(random number) have cdTenantIdList. Then we will have to make more calls, to get the next set of accounts, till we get another 25.

I would suggest for now keep it simple, and use your first method of calling `listAccounts` till the output is not truncated, then process that by filter accounts with `cdTenantIdList` and then applying the limit.

What do you think?

Reply:
It will be simple to implement but will not be cost effective as we will be calling the `listAccounts` API too many times. I would suggest may be computing the count of Accounts with `cdTenantsIdList` each time after `listAccounts` API call and make the next call like 
* If `(limit - cdTenantsIdCount)` > 1000, call `listAccounts` with `max-limit` as 1000
* Otherwise call `listAccounts` with `max-limit` as `(limit - cdTenantsIdCount)`

Reply:
The new listing design with filters is not updated yet.